[KV Transfer] Add optional int8 quantization for KV cache transfer to P2pNcclEngine

xbfs opened the pull request:
## Summary

This PR adds optional int8 quantization support for KV cache transfer in `P2pNcclEngine` to reduce network bandwidth usage and improve transfer throughput, especially in bandwidth-constrained scenarios.

## Motivation

When transferring KV cache across nodes in distributed inference setups, network bandwidth can become a bottleneck. By quantizing tensors from float16/float32 to int8 during transfer, we can reduce the data size by approximately 50-75%, significantly improving transfer speed in scenarios where network bandwidth is limited.

## Changes

- Added `quantize_kv_on_transfer` configuration option (boolean, default: `false`)
- Implemented symmetric int8 quantization with per-tensor scaling
- Added `_quantize_tensor_for_send()` and `_maybe_dequantize_tensor()` helper methods
- Integrated quantization into both PUT and GET transfer paths
- Updated logging to include quantization status

## Technical Details

- **Quantization method**: Symmetric quantization with per-tensor scale factor
  - Formula: `q = clamp(round(x / scale), -128, 127)`
  - Scale calculation: `scale = max(abs(x)) / 127.0`
- **Dequantization**: `x = q * scale`
- **Metadata transfer**: Quantization metadata (scale, original dtype) is transmitted via ZMQ messages
- **Compatibility**: Works with PUT, PUT_ASYNC, and GET transfer modes

## Usage

Enable quantization by setting `quantize_kv_on_transfer: true` in the KV connector extra config:

{
  "kv_connector": "P2pNcclConnector",
  "kv_connector_extra_config": {
    "quantize_kv_on_transfer": true
  }
}

## Benefits

- **Bandwidth reduction**: ~50-75% reduction in transfer data size for float16/float32 tensors
- **Improved throughput**: Faster transfer times in bandwidth-constrained environments
- **Zero breaking changes**: Default behavior unchanged (quantization disabled by default)
- **Backward compatible**: Existing configurations continue to work without modification

## Use Cases

This optimization is particularly beneficial for:
- Cross-node transfers (multi-machine, multi-GPU setups)
- Bandwidth-limited networks (standard Ethernet vs InfiniBand)
- Long sequence lengths (large KV cache sizes)
- High concurrency scenarios (multiple simultaneous transfers)

## Testing

The implementation has been tested with:
- PUT and PUT_ASYNC transfer modes
- Various tensor shapes and dtypes (float16, float32)
- Edge cases (zero tensors, very small/large values)

## Backward Compatibility

- Quantization is **disabled by default** (`quantize_kv_on_transfer: false`)
- No changes to existing APIs or behavior when quantization is disabled
- All existing configurations continue to work without modification

## Future Improvements

Potential enhancements for future PRs:
- Support for different quantization bit-widths (4-bit, etc.)
- Per-channel quantization for better accuracy
- Quantization-aware training integration

Commit:
feat: add optional int8 quantization for KV cache transfer

Add quantize_kv_on_transfer config option to enable int8 quantization during KV cache transfer, reducing network bandwidth usage by ~50-75% for float16/float32 tensors.

- Add quantize_kv_on_transfer config option (boolean, default: false)

- Implement symmetric int8 quantization with per-tensor scaling

- Add quantization/dequantization helper methods

- Fix mypy type checking with proper type casts

Signed-off-by: Bofeng BF1 Xue <xuebf1@Lenovo.com>

robertgshaw2-redhat commented:
@ApostaC - can you be on point for this one?

ApostaC reviewed (commented):
Just some minor comments. Otherwise LGTM!

Commit:
refactor: simplify scale computation logic for better readability

- Remove redundant scale == 0 check in quantization since it's already handled by max_value == 0 condition
- Simplify to single ternary expression for clarity
- Remove unnecessary scale guard in dequantization as scale is guaranteed to be non-zero from quantization

Signed-off-by: Bofeng BF1 Xue <xuebf1@Lenovo.com>

xbfs commented on vllm/distributed/kv_transfer/kv_connector/v1/p2p/p2p_nccl_engine.py:
```diff
@@ -586,6 +599,42 @@ def ping(self):
+        else:
+            scale = float(max_value / 127.0)
+            if scale == 0:
+                scale = 1.0
```
Thank you for the feedback! You're absolutely right. 

The `scale == 0` check on line 614-615 was indeed redundant since the `max_value == 0` condition is already handled above. I've simplified the logic to make it more readable:

**Before:**
if max_value == 0:
    scale = 1.0
else:
    scale = float(max_value / 127.0)
    if scale == 0:
        scale = 1.0
**After:**
scale = float(max_value / 127.0) if max_value > 0 else 1.0

This single-line ternary expression is clearer and eliminates the redundant check. Since we guarantee that `scale` is never 0 in the quantization function, I've also removed the unnecessary guard in the `_maybe_dequantize_tensor` function for consistency.

The changes have been pushed. Thanks again for helping improve the code quality! 🙏

Commit:
style: add newline at end of file

Signed-off-by: Bofeng BF1 Xue <xuebf1@Lenovo.com>

Commit:
refactor: optimize quantization code and fix tensor device placement

- Remove redundant numel() check in quantization (max() always returns scalar)

- Merge abs_max calculation into single line for cleaner code

- Fix tensor device placement: move to(device) after quantization to avoid unnecessary transfers

Signed-off-by: Bofeng BF1 Xue <xuebf1@Lenovo.com>

xbfs committed:
Merge branch 'main' into feat/kv-transfer-quantization

xbfs commented:
Hi @ApostaC,
May I know if anything else is needed for this PR to be merged?  
I'd be happy to update it.

xbfs committed:
Merge branch 'main' into feat/kv-transfer-quantization
